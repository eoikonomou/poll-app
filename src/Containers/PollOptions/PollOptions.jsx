import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { observer } from 'mobx-react';
import InputField from '../../Components/InputField/InputField';
import { generateUUID, deepCopy } from '../../../utils/generalUtils';
import './PollOptions.scss';

const PollOptions = observer(({ store }) => {
  const [newOption, setNewOption] = useState('');
  const {
    pollQuestion,
    pollOptions,
    setPollQuestion,
    setPollOptions,
    addPollOption,
    reset,
    pollLocked,
    setPollLocked,
  } = store;

  const onPollQuestionChange = (e) => {
    setPollQuestion(e.target.value || '');
  };

  const onOptionChange = (optionId) => (e) => {
    const optionIndex = pollOptions.findIndex((option) => option.id === optionId);
    setPollOptions([
      ...pollOptions.slice(0, optionIndex),
      { id: optionId, value: e.target.value },
      ...pollOptions.slice(optionIndex + 1),
    ]);
  };

  const onAddNewOption = () => {
    addPollOption({ id: generateUUID(), value: newOption, count: 0 });
    setNewOption('');
  };

  const onDeleteOption = (optionId) => () => {
    const optionsCopy = deepCopy(pollOptions);
    const optionIndex = optionsCopy.findIndex((opt) => opt.id === optionId);
    optionsCopy.splice(optionIndex, 1);
    setPollOptions(optionsCopy);
  };

  const resetPoll = () => {
    reset();
    setNewOption('');
  };

  const onNewOptionChange = (e) => {
    setNewOption(e.target.value || '');
  };

  const lockPoll = () => {
    setPollLocked(true);
  };

  return (
    <div className="section space">
      <div className="main">
        <InputField
          id="poll-question"
          value={pollQuestion}
          onChange={onPollQuestionChange}
          disabled={pollLocked}
        />
        <button
          type="button"
          disabled={pollOptions.length < 2 || pollQuestion.length === 0}
          onClick={lockPoll}
          style={{ verticalAlign: 'middle' }}
          id="lock-btn"
          className={pollLocked ? 'locked' : ''}
        >
          <i className="material-icons" style={{ fontSize: '26px  ' }}>{pollLocked ? 'lock' : 'lock_open'}</i>
        </button>
        <div className="flex-container">
          <>
            {pollOptions.map((option) => (
              <div className="option-container" key={option.id}>
                <InputField
                  id={option.id}
                  onChange={onOptionChange(option.id)}
                  value={option.value}
                  placeholder="Add new option"
                  disabled={pollLocked}
                />
                <button
                  type="button"
                  className="btn del-btn"
                  onClick={onDeleteOption(option.id)}
                  disabled={pollLocked}
                >
                  X
                </button>
              </div>
            ))}
            {(pollOptions.length < 10 && !pollLocked) && (
              <div className="option-container" key="new_option_container">
                <InputField
                  id="new_option"
                  onChange={onNewOptionChange}
                  value={newOption}
                  placeholder="Add new option"
                  disabled={pollLocked}
                />
                <button
                  type="button"
                  id="add_new_option"
                  className="btn add-btn"
                  onClick={onAddNewOption}
                  disabled={
                    pollOptions.find((option) => option.value === newOption)
                    || newOption.length === 0
                  }
                >
                  Add
                </button>
                {pollOptions.find((option) => option.value === newOption) && (
                  <div className="error-msg">Duplicate options are invalid</div>
                )}
              </div>
            )}
          </>
        </div>
      </div>
      <div className="footer">
        <div className="message">
          {`${pollOptions.length}/10 possible answers`}
        </div>
        <div className="btn-holder">
          <button type="button" onClick={resetPoll} className="btn reset-btn">Reset</button>
        </div>
      </div>
    </div>
  );
});

PollOptions.displayName = 'PollOptions';

PollOptions.propTypes = {
  store: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.array,
    PropTypes.bool,
  ])).isRequired,
};

export default PollOptions;
