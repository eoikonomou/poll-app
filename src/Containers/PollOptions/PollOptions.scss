.option-container {
  margin-bottom: 15px;

  .material-icons {
    font-size: 22px;
    vertical-align: middle;
    color: #ad0000;
    cursor: default;
  }

  .error-msg {
    color: red;
  }
}

#poll-question {
  height: 32px;
  margin-bottom: 5px;
}

.locked {
  background-color: '#EED179';
}

.footer {
  display: flex;
  flex-direction: row;
  font-weight: 600;

  .message {
    flex: 1;
  }

  .btn-holder {
    margin-right: 5px;
  }
}
