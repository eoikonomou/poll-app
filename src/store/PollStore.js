import { action, observable } from 'mobx';

export default class PollStore {
  @observable pollQuestion = '';

  @observable pollOptions = [];

  @observable pollLocked = false;

  @action setPollQuestion = (data) => {
    this.pollQuestion = data;
  }

  @action setPollOptions = (data) => {
    this.pollOptions = data;
  }

  @action addPollOption = (data) => {
    this.pollOptions.push(data);
  }

  @action addVote = (optionId) => {
    const option = this.pollOptions.find((opt) => opt.id === optionId);
    option.count += 1;
  }

  @action reset = () => {
    this.pollOptions = [];
    this.pollQuestion = '';
    this.pollLocked = false;
  }

  @action setPollLocked = (data) => {
    this.pollLocked = data;
  }
}
