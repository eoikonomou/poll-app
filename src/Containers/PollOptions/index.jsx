import PollOptions from './PollOptions';

export default PollOptions;
