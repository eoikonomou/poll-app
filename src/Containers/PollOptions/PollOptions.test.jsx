import React from 'react';
import { shallow, mount } from 'enzyme';
import PollOptions from './PollOptions';
import PollStore from '../../store/PollStore';

describe('PollOptions', () => {
  let store = null;

  beforeEach(() => {
    store = new PollStore();
  });

  afterEach(() => {
    store = null;
  });

  test('should render successfully', () => {
    const wrapper = shallow(<PollOptions store={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should change pollQuestion when #poll-question input value changes', () => {
    const wrapper = mount(<PollOptions store={store} />);
    const pollQuestionInput = () => wrapper.find('InputField').at(0);
    pollQuestionInput().simulate('change', { target: { value: 'test-question' } });
    expect(store.pollQuestion).toEqual('test-question');
  });

  test('should have lock-btn disabled while pollOptions are less than 2 and pollQuestion is empty', () => {
    let wrapper = shallow(<PollOptions store={store} />);
    const lockButton = () => wrapper.find('#lock-btn').at(0);
    expect(lockButton().prop('disabled')).toBe(true);
    store.setPollOptions([
      { id: 'opt1', value: 'opt1', count: 0 },
      { id: 'opt2', value: 'opt2', count: 0 },
    ]);
    wrapper = shallow(<PollOptions store={store} />);
    expect(lockButton().prop('disabled')).toBe(true);
    store.setPollQuestion('test');
    wrapper = shallow(<PollOptions store={store} />);
    expect(lockButton().prop('disabled')).toBe(false);
  });
});
