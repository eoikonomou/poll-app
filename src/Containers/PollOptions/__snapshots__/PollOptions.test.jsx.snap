// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollOptions should render successfully 1`] = `
<div
  className="section space"
>
  <div
    className="main"
  >
    <InputField
      disabled={false}
      id="poll-question"
      onChange={[Function]}
      placeholder="Input the question here..."
      value=""
    />
    <button
      className=""
      disabled={true}
      id="lock-btn"
      onClick={[Function]}
      style={
        Object {
          "verticalAlign": "middle",
        }
      }
      type="button"
    >
      <i
        className="material-icons"
        style={
          Object {
            "fontSize": "26px  ",
          }
        }
      >
        lock_open
      </i>
    </button>
    <div
      className="flex-container"
    >
      <div
        className="option-container"
        key="new_option_container"
      >
        <InputField
          disabled={false}
          id="new_option"
          onChange={[Function]}
          placeholder="Add new option"
          value=""
        />
        <button
          className="btn add-btn"
          disabled={true}
          id="add_new_option"
          onClick={[Function]}
          type="button"
        >
          Add
        </button>
      </div>
    </div>
  </div>
  <div
    className="footer"
  >
    <div
      className="message"
    >
      0/10 possible answers
    </div>
    <div
      className="btn-holder"
    >
      <button
        className="btn reset-btn"
        onClick={[Function]}
        type="button"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
